### PR TITLE
Run our tests against Node.js 20, 22, and 23

### PR DESCRIPTION
**Pull request by jasonaowen**
We are currently targeting Node.js 20, which is in maintenance LTS. 22 is the active LTS, and 23 is for current development.

Run our tests against all of these, so that we can be ready for new versions.

### CODE REVIEW

**Comment by jasonaowen** [comment]
We'll need to update our branch protection rules again. Probably we only need to require `tests (20)` rather than all three, but if folks have opinions I'm happy to defer to the group!

**Comment by slifty** [comment]
A series of thoughts!

This will triple the CI time right?  Or are they run in parallel.

I'm not entirely sure of the value of testing against versions of the node we don't have a policy of supporting, as opposed to simply updating to support them at the time that we determine we support them.  This is especially true for versions we explicitly never intend to support by policy (odd numbered versions)

If it fails on, for instance, 23 would the dev be expected to make changes to support 23?  If not, what's the point of running the test?

I think we should update our node versioning policy MD file to reflect our intentions if we decide to merge this change.

**Comment by jasonaowen** [comment]
> This will triple the CI time right? Or are they run in parallel.

They run in parallel!

> I'm not entirely sure of the value of testing against versions of the node we don't [have a policy of supporting](https://github.com/PhilanthropyDataCommons/service/?tab=readme-ov-file#node-version), as opposed to simply updating to support them at the time that we determine we support them according to our policy. This is especially true for versions we explicitly never intend to support by policy (odd numbered versions)

I think the benefit is knowing that we could move to a new version when we want, rather than hoping that such an eventual migration will be uneventful.

In particular, I wanted to pave the way for upgrading to 22, prompted by https://github.com/PhilanthropyDataCommons/service/pull/1259#discussion_r1868063298 .

> If it fails on, for instance, 23 would the dev be expected to make changes to support 23? If not, what's the point of running the test?

It depends on the nature of the failure, I suppose. 24 will fork off of 23 in April and become the new active LTS in October ([release schedule](https://github.com/nodejs/Release/blob/main/README.md#release-schedule)), so keeping an eye on it seems prudent, even if we don't let test failures block merging.

> I think we should update our node versioning policy MD file to reflect our intentions if we decide to merge this change.

Now that I look, we're actually not up to date with our versioning policy!

>> We aim to use the "Active LTS" version of node.

20 is the maintenance LTS version, and 22 is the active LTS version.

So, if we do merge this, once we upgrade we can drop support for 20.

**Comment by slifty** [comment]
> hoping that such an eventual migration will be uneventful.

To me the point of the policy is so that we aren't burdened with having to support versions that we haven't commited to supporting yet (which will either add some % of continuous overhead in the moment).  This PR would mean we, functionally, are supporting those versions.  If there is ever a failure for an unsupported version we'll either be forced to (1) have a red X for every future PR until we fix it, making noise and also making the test meaningless until resolved for that or (2) add in support in that moment.

For instance, say a future version drops support for a feature we are using or changes an API somehow in a breaking way -- yes that will represent a technical effort at some point, but the point of our policy is that we can prioritize that upgrade effort rather than coupling it with a given feature PR.

Also regarding the dev version, I understand it's forked from a dev version but isn't the point of the dev version that things might change more rapidly / experimentally / there may be more compatibility issues than the ultimate forked version is expected to have?  What's the value in providing continuous support for a version that is potentially in flux.

**Review by bickelj — APPROVED** [review]
I think we should let the `node` version in the `package.json` match the same test matrix but otherwise this looks great.

**Review by bickelj — CHANGES_REQUESTED** [review]
While trying to figure out how DO chooses a node version, it appears [DO looks at the `package.json`](https://docs.digitalocean.com/products/app-platform/reference/buildpacks/nodejs/#specify-a-nodejs-engine), in which case this change might change the version of node used on production deployments, in which case I hesitate to approve. I want the tests to run on all those versions but I want the exact version of node for production declared. Failing the ability to declare an exact version for deployment, at least the major version of node should be controlled for production deployment.

File: package.json
Comment (slifty) [reply package.json:44]: Truly if we are testing against future node versions in our CI (as opposed to, say, a monthly cron action that runs tests against future versions) then we are, functionally, supporting them.

That is fine, but it should be framed in that way and we should update the policy to say "this project commits to supporting the active and development releases of node"

File: .github/workflows/test.yml
Comment (slifty) [reply .github/workflows/test.yml:13]: @bickelj I suppose it comes down to this: if the dev version has churn before the fork, do we want to be continuously making sure we support all versions of the churn.

**Comment by slifty** [comment]
I'll just open the PR to update us to be in line with our policy which will help narrow the scope of this discussion to "should we support the dev version of node"

File: .github/workflows/test.yml
Comment (slifty) [reply .github/workflows/test.yml:13]: > I want to take it even one step further and add the exact expected production node version.

This would mean we don't benefit from the latest security patches without a manual developer action.  IMHO let's stick with the node ecosystem's general standard of major version being the level that we specify for test / deploy.

Comment (bickelj) [reply .github/workflows/test.yml:13]: Your premise is incorrect, though. We lack the ability to get the latest security patches even if we wanted them when using DO's buildpack. We're stuck on an old minor version. We can do better than the general node population here : )

**Comment by slifty** [comment]
I also made an issue to capture the idea of updating the version policy (and practice) so that we're testing against `Current` instead of just `Active` which I do appreciate has merit!

**Pull request closed without merging**
